### ISSUE TITLE
Prisma unable to connect to Supabase database

### ISSUE THREAD

**Issue by DeviousLab** [supabase/supabase#7772]
# Bug report

## Describe the bug

Prisma npx commands and Prisma Studio are unable to connect to the Supabase database. 

Running `npx prisma migrate dev` yields in the following error message:
```
> npx prisma migrate dev
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": PostgreSQL database "postgres", schema "public" at "db.xltfmrlaegzuuopvypxx.supabase.co:5432"

Error: P1001: Can't reach database server at `db.xltfmrlaegzuuopvypxx.supabase.co`:`5432`

Please make sure your database server is running at `db.xltfmrlaegzuuopvypxx.supabase.co`:`5432`.
```
I followed the advice at https://github.com/prisma/prisma/issues/13807 and I was indeed able to migrate an initial db after creating a temporary table. 

But after running Prisma Studio I am unable to add any entries and once more unable to migrate a database:
 
![image](https://user-images.githubusercontent.com/53790060/179614060-173a70b6-55cd-4df7-a610-085d988b445b.png)
```
Type: undefined
Message: 
Invalid `prisma.home.create()` invocation:


  Can't reach database server at `db.xltfmrlaegzuuopvypxx.supabase.co`:`5432`

Please make sure your database server is running at `db.xltfmrlaegzuuopvypxx.supabase.co`:`5432`.

Code: undefined

Query:
[object Object]
```
I made sure that my database password contains no special characters as #4161 discussed. 

Using psql to connect to the data base shows no issues either and shows the appropriate tables:
```
> psql -h db.xltfmrlaegzuuopvypxx.supabase.co -p 5432 -U postgres -d postgres
Password for user postgres: 
psql (13.4, server 14.1)
WARNING: psql major version 13, server major version 14.
         Some psql features might not work.
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
Type "help" for help.

postgres=> \c postgres
psql (13.4, server 14.1)
WARNING: psql major version 13, server major version 14.
         Some psql features might not work.
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
You are now connected to database "postgres" as user "postgres".
postgres=> \dt
               List of relations
 Schema |        Name        | Type  |  Owner   
--------+--------------------+-------+----------
 public | Home               | table | postgres
 public | _prisma_migrations | table | postgres
(2 rows)
```
![image](https://user-images.githubusercontent.com/53790060/179614727-99f38a3a-8d31-4981-8e26-2214a2848315.png)

## System information

- OS: Fedora 35
- Browser: Brave v1.41.96
- Version of Node.js: [16.16.0]

## Additional context

Prisma schema:
```
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Home {
  id          String   @id @default(cuid())
  image       String?
  title       String
  description String
  price       Float
  guests      Int
  beds        Int
  baths       Int
  createdAt   DateTime @default(now())
  updatedAt   DateTime @updatedAt
}
```

**Comment by kristianeboe** [comment]
Prisma did release a  4.0 not long ago, but I have not upgraded yet...

**Comment by kiwicopple** [comment]
could you please share which version you're running @kristianeboe ?

**Comment by kristianeboe** [comment]
"@prisma/client": "^3.15.2",

**Comment by DeviousLab** [comment]
> Perhaps there was a change on the Prisma side? Did you also open this issue with the Prisma team? I'll also give them a ping and see if they have any ideas

I was unsure whether to bring it up on the Prisma or Supabase repo, but since I sometimes faced loading the database even with the Supbase dashboard, I figured it could be an issue best brought up here.

> Hope this helps :) I will soon upgrade to Prisma 4 and see if that fixes it too.

My project is currently using @prisma/client": "^4.0.0", while facing these errors.

**Comment by kiwicopple** [comment]
>since I sometimes faced issues loading the database even with the Supbase dashboard

From the prisma side, is this intermittent, or are you always getting the error?

If you can share more about the Dashboard , would you mind opening another github issue? That's definitely something we want to solve, but I don't want to conflate the two here

**Comment by kristianeboe** [comment]
I'm on WSL 2 on Windows, just regular Ubuntu

**Comment by nikpl7777** [comment]
Cannot setup direct connection to database neither from PgAdmin not NodeJS server. Getting the same error.

**Comment by kristianeboe** [comment]
My issue was solved by working from home hehe.
When I had my issue I was trying to connect from a hotel lobby, but from my own apartment it seems to work fine.

**Comment by janpio** [comment]
Then it was probably a timeout error and adding a parameter `connect_timeout` to the connection string would have helped as well @kristianeboe with a higher than default value.

**Comment by kristianeboe** [comment]
Thank you, that's great to know :D

**Comment by noyan-alimov** [comment]
i am having the same issue using a hotel wifi, but it worked fine at home.
if this info helps anyone :)

**Comment by code-foracademic** [comment]
I am also having the same issue. The problem seems to be that it is only allowing connection to the database with the initial wifi that was used to set it up. Whenever I connect to another wifi network, I'll get the same issue.

**Comment by WHenderson** [comment]
I'm having the same problems.

Running `prisma migrate dev --name init`

Using the connection pooling string gives the error 
* `Error: db error: FATAL: unexpected response from login query`

Using the regular connection string gives these errors 
* `Error: P1001: Can't reach database server at`
* `Error: P1011: Error opening a TLS connection: unexpected EOF during handshake`

I'm using prisma v4.6.1

**Comment by egor-romanov** [comment]
Hey @WHenderson, I believe this should be another issue from Prisma side. Prisma tries to drop the whole DB, which would lead to the corruption of supabase managed services.
We updated our integration guide to respect this prisma behaviour: https://supabase.com/docs/guides/integrations/prisma
Please take a look at the section with "Prisma Shadow Databases"

**Comment by janpio** [comment]
Can you help us at Prisma figure out which resources we are removing that we should not? We are more than happy to adapt Prisma's behavior if that is possible.

**Comment by egor-romanov** [comment]
Hey @janpio,
Wow, that would be awesome!

Running the `prisma migrate dev` command will drop the current database and remove all of the migrations, schemas, extensions, roles, and policies that are managed by Supabase. This could have serious consequences for our services. 

It would be preferable if Prisma could avoid dropping the database when running the `prisma migrate dev` command.

However, there still may be conflicts with Prisma's migrations that involve the installation of extensions, as we have our own process for installing extensions.

**Comment by janpio** [comment]
Prisma "only" runs SQL commands, that users could also run by hand. Does Supabase only support installing extensions via the UI?

> It would be preferable if Prisma could avoid dropping the database when running the `prisma migrate dev` command.

`migrate dev` itself never deletes any data unless that is part of the migration. If there is drift detected, so basically the database and our migration history do not match, we might suggest a "reset" to a user - and that then tries to delete the objects present to get the database into a clean state. If that fails, it will get the big hammer and actually drop the database and recreate it.

Do you know which of these two modes are causing the problems?
Is there a list of the items should _not_ be removed under any circumstances?

**Comment by egor-romanov** [comment]
Supabase installs a lot of extensions, and we know that some of our users have had conflicts with Prisma when trying to install extensions that were already installed. However, this is a rare occurrence and we don't have much feedback to share on it. 

> If that fails, it will get the big hammer and actually drop the database and recreate it.
This is what usually happens. Supabase offers not only a Postgres instance, but also a lot of other services that rely on their own schemas in the database. And when Prisma tries to drop the database, it can cause issues with the user's project. The list of items that could be affected is extensive and includes the removal of any data, schemas, roles, or extensions.

To summarize, it seems that conflicts can be avoided if Prisma only makes changes to the `public` schema and does not include a 'drop schema' statement. This would allow Supabase's services and the user's project to function properly.

**Comment by janpio** [comment]
Yeah, we only do that if we _need_ to - so just not doing that is not really an options unfortunately. I'll check why and when we do the hammer method instead of nicely deleting the individual items we need to get rid of.

**Comment by vuolo** [comment]
> My issue was solved by working from home hehe. When I had my issue I was trying to connect from a hotel lobby, but from my own apartment it seems to work fine.

It was my VPN! How simple. Thank you so much. Spent a little too much time on this one... lol

**Comment by WHenderson** [comment]
Ironically, I need to connect to a VPN to solve my issue. lol

**Comment by developerGCGMMC** [comment]
Add connection timeout at the end of your DATABASE_URL.

?connect_timeout=300

**Comment by saidMounaim** [comment]
> Add connection timeout at the end of your DATABASE_URL.
> 
> ?connect_timeout=300

Thanks! it's working for me

**Comment by ludbek** [comment]
Try creating db in another region.

**Comment by bottercode** [comment]
> > Cannot setup direct connection to database neither from PgAdmin not NodeJS server. Getting the same error.
> 
> Did you got any lead to connect the pgadmin with supabase connection info

works well if you add a ?connect_timeout=300. Adds a timeout to the string which makes the database responsive then.
Guess this issue is with the Prisma. If anyone finds it can you link one form me?

**Comment by overmode** [comment]
Also make sure that special characters are escaped in your password, otherwise it will break :
Character   Percent-Encoded
----------  -----------------
:          %3A
/          %2F
?          %3F
&          %26
=          %3D
@          %40
+          %2B
%          %25
#          %23

**Comment by jynzo94** [comment]
I had this issue because I used VPN. I turned it off and it works. I waste many hours in debug...

**Comment by YugBhanushali** [comment]
Had the same issue and found out that the supabase provide IPv6 address only with free tier and my wifi only had the IPv4 address so I had to configure the IPv6 address manually to my wifi setting.

You can get the IPv6 address from your ISP.

**Comment by balqaasem** [comment]
Mine was solved by removing the square braces around the PASSWORD

**Comment by developer-256** [comment]
I was having the same problem as you and I found a solution.

_

The **DATABASE_URL** should be **Transaction pooler** string.

And **DIRECT_URL** should be **Session pooler** string.

_

This is shown here: https://github.com/prisma/prisma/issues/11643#issuecomment-2315943032

**Comment by 1expel** [comment]
I have been using the IPv6 the entire time and it has been working. All of a sudden I got this error and switching to IPv4 connection via Session Pooler seemed to work for me.
